radicle-node/src/tests.rs

```diff
@@ -593,6 +593,52 @@ fn test_refs_announcement_no_subscribe() {
     assert!(alice.messages(eve.id()).next().is_none());
 }
 
+#[test]
+fn test_queued_fetch() {
+    let mut alice = Peer::new("alice", [7, 7, 7, 7]);
+    let bob = Peer::new("bob", [8, 8, 8, 8]);
+    let rid = arbitrary::gen::<Id>(1);
+    let rid2 = arbitrary::gen::<Id>(1);
+    let (send, _recv) = chan::bounded::<node::FetchResult>(1);
+
+    logger::init(log::Level::Debug);
+
+    // Send the first fetch.
+    alice.connect_to(&bob);
+    alice.command(Command::Fetch(rid, bob.id, send));
+
+    assert_matches!(alice.messages(bob.id).next(), Some(Message::Fetch { .. }));
+
+    // Send the 2nd fetch that will be queued.
+    let (send2, _recv2) = chan::bounded::<node::FetchResult>(1);
+    alice.command(Command::Fetch(rid2, bob.id, send2));
+
+    // We shouldn't send out the 2nd fetch to Bob while we're doing the 1st fetch.
+    assert_matches!(alice.messages(bob.id).next(), None);
+
+    alice.receive(bob.id(), Message::FetchOk { rid });
+    assert_matches!(alice.messages(bob.id).next(), None);
+
+    // Have enough time pass that Alice sends a "ping" to Bob.
+    alice.elapse(KEEP_ALIVE_DELTA);
+
+    // Finish the 1st fetch.
+    alice.fetched(
+        Fetch {
+            rid,
+            namespaces: Namespaces::All,
+            remote: bob.id,
+            initiated: true,
+        },
+        Ok(vec![]),
+    );
+
+    // Now the 1st fetch is done, the messages Bob missed should be relayed to him.
+    let mut messages = alice.messages(bob.id);
+    assert_matches!(messages.next(), Some(Message::Fetch { .. }));
+    assert_matches!(messages.next(), Some(Message::Ping { .. }));
+}
```

Comment:
Such a beautiful test :smiling_face_with_tear: 

I think we're missing an important detail though! We are checking that alice is indeed sending the queued `Fetch` message to Bob, but not that she's actually ready to fetch from Bob the second time (or the first time, but that's tested elsewhere I guess).

What we should do at the end of this test is additionally:
1. Have alice receive the `FetchOk` from bob
2. Expect that alice outputs an `Io::Fetch` action (not a message!), which is how the service gears of for fetching. You can get these via `alice.outbox()`.

My guess is that this will cause test failure:
```rust
            (session::State::Connected { protocol, .. }, Message::FetchOk { rid }) => {
                if *protocol
                    != (session::Protocol::Gossip {
                        requested: Some(rid),
                    })
                {
                    [...]
                    return Err(session::Error::Misbehavior);
                }

```

Reply:
thanks for the details!